### PR TITLE
fix(masthead): apply global styles updates based on Unified Cloud Masthead project

### PR DESCRIPTION
**Pull request by proeung**
### Related Ticket(s)

DDS Ticket: N/A
HC JIRA Ticket: https://jsw.ibm.com/browse/HC-1679

### Description
- This PR implements new design enhancements to the dotcom and Cloud masthead that came out of the "Unified Cloud Masthead + Mega Menu Project" (see project overview doc - https://ibm.ent.box.com/notes/775640792107).
- We're using this PR to capture the overall design QA for the Cloud masthead + mega menu front-end contribution effort.
- Expect a fail in the Percy visual diffs as these changes are applied to the dotcom masthead container as well.

### Design Spec
- Masthead: https://ibm.ent.box.com/file/786859527229
- Mega Menu: https://ibm.ent.box.com/file/774092317795

### Changelog

Cloud Masthead
Screen Shot 2021-04-28 at 10 19 06 AM

Dotcom Masthead
Screen Shot 2021-04-28 at 10 18 54 AM

- All of the design changes in this PR are based on the specs and docs that the design team has agreed upon (see doc - https://ibm.ent.box.com/notes/787865510080).

**Global Changes** 

- [x] All L0 menu item text reduced to 14px (from 16px) now using $body-short-01 
- [x] All drop down menu item text sizes reduced to 14px (from 16px) now using $body-short-01 
- [x] All L0 menu icons reduced to 16px (from 20px)
- [x] All L0 menu items using $text-02 for initial state (changes to $text-01 on hover and selected states)
- [x] All drop down menu items using $text-02 for initial state (changes to $text-01 on hover and selected states)
- [x] "Log in" text for logged out state of the menu (no drop down in logged out state)
- [x] User icon for account drop down once logged in (menu items will vary between IBM.com and Cloud)

**Testing Instructions**
- Visit the `Cloud Masthead` Storybook variant - `/?path=/story/components-cloud-masthead--default`
- Ensure that the masthead + mega menu components matche with the design specs for all breakpoints.
- Visit the `Dotcom Masthead` Storybook variant - `/?path=/story/components-masthead--default`
- Ensure that the masthead + mega menu components matche with the design specs for all breakpoints.

### CODE REVIEW

**Comment by huntermacd** [comment]
I've noticed the follow discrepancies between the preview and the design for the Cloud masthead:

- :active state incorrect, should be blue underline only, not blue outline

Design:
Screen Shot 2021-04-28 at 12 22 29 PM

Preview:
Screen Shot 2021-04-28 at 12 23 16 PM

- drop-down menu item divider should not extend all the way

Design:
Screen Shot 2021-04-28 at 12 26 43 PM

Preview:
Screen Shot 2021-04-28 at 12 27 27 PM

- “cost estimator” listed twice under Pricing
- search icon height/width not equal
- open search bar should have 32px padding between it and Cloud name
- horizontal divider should not extend all the way in search
- horizontal dividers should exist in user menu
- carousel overflow should not be visible below 960px
- menu items should move to hamburger menu below 960px
- search results should extend full width at 320px

Not sure which of these fall under the purview of this ticket.

**Review by shixiedesign — CHANGES_REQUESTED** [review]
Still seeing Dotcom masthead showing to 14px. This is not approved. This PR should have typestyle changes **contained within Cloud masthead**.

**Comment by jeffchew** [comment]
@proeung could the global changes be scoped to cloud masthead for now? I think we need more time to look through and explore this change, and hopefully we can do in a separate PR once we have final design direction.

**Comment by proeung** [comment]
@jeffchew @shixiedesign @ltitle Let's take time from the "Cloud unified coordination - dev + strategy" meeting tomorrow to discuss how to best move forward with this. I was under the impression that the `14px` font size is something that we'd like to implement globally based on the conversations within this ([slack thread](https://ibm-marketing-comms.slack.com/archives/G01LKRP5RB6/p1616520955027500?thread_ts=1615918282.006200&cid=G01LKRP5RB6)), however, if that's not the case then we'll need to discuss the implications of having duplicate styles and components in order to scope some of these design updates within the `dds-cloud-masthead-container`. Happy to talk more tomorrow. 👍

**Comment by annawen1** [comment]
> I've noticed the follow discrepancies between the preview and the design for the Cloud masthead:
> 
> * :active state incorrect, should be blue underline only, not blue outline
> 
> Design:
> Screen Shot 2021-04-28 at 12 22 29 PM
> 
> Preview:
> Screen Shot 2021-04-28 at 12 23 16 PM

The `active` state in the designs look like our `selected` state (where the menu item has the selected states on load) in our current Masthead with the blue underline...maybe it is supposed to be `selected` state instead?

File: packages/styles/scss/components/masthead/_masthead-search.scss
Comment (kennylam) [inline packages/styles/scss/components/masthead/_masthead-search.scss:92]: Thanks for this @proeung! This removes the 1px horizontal rule between the logo and the platform name. Is that intended? Because I don't see it in the design specs.

image

**Comment by ltitle** [comment]
### **Masthead Design QA done in Chrome:**

Selected state for items in the menu (when they have the blue bar underneath) should use color: $text-01 for the menu item text. It looks like it is currently using $text-02 in react experimental 
image

Icons in masthead should be using $icon-02 color until hovered or selected. As seen in react experimental
image

Contact us and Log in text for Cloud masthead should be using color: $text-02 until hovered or selected as seen in web components experimental
image

X to close search is not showing up on small breakpoint when search bar is opened. As seen in web components experimental.
image

Menu items and chevrons in the hamburger nav should use color: $text-02 for items until they are hovered or selected. As seen in web components experimental: 
image

I could not figure out how to test the logged in state in storybook without it taking me out of the experience

Menu items in hamburger should never be semi-bold weight. Keep all items at regular weight just changing text color when selected. As seen in react experimental. 
image

**Comment by ltitle** [comment]
### **Mega menu design QA done in Chrome, web components experimental storybook**

Left side items in mega menu should use color:$text-01 for hover and selected state. As seen in web components experimental.
image

If the menu items on the right start a new row the next item should always right align. Currently seen mixing center and right aligned items as well. As seen in web components experimental 
image
image
Fill across on all rows before starting a new one: 
image

Arrow next to Workforce Management should always stay on the 1st line and not wrap. 
image

Mega menu should have a 320px minimum height. For Solutions menu the height is also changing depending on which tab on the left is selected. The height should be determined by the tab with the longest content so that the height within a single mega menu will not change with each click. It appears to be working this way on the Products mega menu but not for Solutions.
image

If the left side menu tabs do not fit vertically in the browser window I would expect that we have scrolling only for that side of the mega menu rather than making both sides scroll. Once the content does not fit I expect 2 scroll bars one for left and one for right side content as needed. 
image
Reference spec w 2 scroll bars:
image

White gradient is missing for content that needs to scroll. View all products is also intended to maintain sticky to the bottom even when there is scrolling
image
See reference specs:
image

Should not have a darker full bleed rule after Cloud Pak Solutions 
image

I believe we decided that the mega menu for products would only go down 2 levels on mobile so we will only go down to the categories rather than list of product names. Can confirm w Keegan but reference sceenshot from specs 
image

I think it would make sense if the View all products keeps the blue styling it had at larger breakpoints even at mobile size 
image

**Comment by proeung** [comment]
Hey, @ltitle! The majority of your design QA feedback should be addressed in the latest commits that I pushed up, however, there are a few comments that I still have questions about. Please take a look below and let me know if I missed anything. I'd like to get all of these changes resolved and tested so that these design updates will go out as part of the **June 15th** release, which has a code freeze of **June 4th.**

**Web Components preview:** (You should see the latest changes once the CI has finished building.)

- Cloud Masthead + Megamenu: https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5922/index.html?path=/story/components-cloud-masthead--default
- Default Masthead + Megamenu: https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5922/index.html?path=/story/components-masthead--default

------------------------

> X to close search is not showing up on small breakpoint when search bar is opened. As seen in web components experimental.

I'm not able to replicate this issue that you're seeing within the mobile breakpoints. Can you let me know the specific browser size you're seeing this issue in?

>If the left side menu tabs do not fit vertically in the browser window I would expect that we have scrolling only for that side of the mega menu rather than making both sides scroll. Once the content does not fit I expect 2 scroll bars one for left and one for right side content as needed.

What you're seeing is coming from the global masthead scroll bars, I need to touch base with @annawen1 and see what's the best way to accomplish this change without affecting the global version.

>Should not have a darker full bleed rule after Cloud Pak Solutions

I just pushed up a change within the Cloud Masthead JSON endpoint that removes this dark highlighted line (see - https://github.ibm.com/webstandards/ibm-dotcom-translations/pull/375).

>I think it would make sense if the View all products keeps the blue styling it had at larger breakpoints even at mobile size

If we want to change the color of this "View all" link to blue, then we'll need to do this on the global side as well, since this CTA shared the same component.

File: packages/styles/scss/components/masthead/_masthead-search.scss
Comment (proeung) [reply packages/styles/scss/components/masthead/_masthead-search.scss:92]: See my comment here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5922#issuecomment-837454977

**Review by annawen1 — APPROVED** [review]
Looks good to me! Thanks for making all those changes!

**Comment by proeung** [comment]
@annawen1 Thanks for the code review!

@ltitle @shixiedesign Can get another round of design QA from one of you? I'd like to get all of these changes resolved and tested so that these design updates will go out as part of the June 15th release, which has a code freeze of 06/04 and a code merge/submission deadline of **06/02**.

**Comment by ltitle** [comment]
X to close search is not showing up on small breakpoint when search bar is opened. As seen in web components experimental.
I'm not able to replicate this issue that you're seeing within the mobile breakpoints. Can you let me know the specific browser size you're seeing this issue in?

**It looks like this is happening at 670px and smaller. I also noticed that we are not breaking to the hamburger until around 800px. I think we should bread at 959px instead since the menu items are unusable at that size.**

----

**I am also noticing a new issue where once you use the menu horizontal scroll to paginate there is a 16px gap to the left of the Cloud label where the other menu items are peaking through. We should not be able to see that there. It should be white. It am seeing that for the entire time the menu items are scrollable**

image

-----

If the left side menu tabs do not fit vertically in the browser window I would expect that we have scrolling only for that side of the mega menu rather than making both sides scroll. Once the content does not fit I expect 2 scroll bars one for left and one for right side content as needed.
What you're seeing is coming from the global masthead scroll bars, I need to touch base with @annawen1 and see what's the best way to accomplish this change without affecting the global version.

**OK I am still seeing this issue.**

----

Should not have a darker full bleed rule after Cloud Pak Solutions
I just pushed up a change within the Cloud Masthead JSON endpoint that removes this dark highlighted line (see - https://github.ibm.com/webstandards/ibm-dotcom-translations/pull/375).

**I am still seeing this darker line under Analytics** 

image

-------

I think it would make sense if the View all products keeps the blue styling it had at larger breakpoints even at mobile size

If we want to change the color of this "View all" link to blue, then we'll need to do this on the global side as well, since this CTA shared the same component.

**Will confirm this with @shixiedesign and @oliviaflory**

**Comment by proeung** [comment]
@ltitle Thanks for doing another round of design QA and answering all of my questions from [this comment](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5922#issuecomment-846090012). Overall, I was able to address some of the usability issues that you've noted, however, there are a couple that we might want to revisit in another ticket. For now, I think the changes here are good for merging down. Let me know if you feel the same way as well.

>It looks like this is happening at 670px and smaller. I also noticed that we are not breaking to the hamburger until around 800px. I think we should break at 959px instead since the menu items are unusable at that size.

I pushed a commit that should resolve the display issue you're seeing within the search bar. See latest build preview link - https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5922/index.html?path=/story/components-cloud-masthead--default

>I am also noticing a new issue where once you use the menu horizontal scroll to paginate there is a 16px gap to the left of the Cloud label where the other menu items are peeking through. We should not be able to see that there. It should be white. I am seeing that for the entire time the menu items are scrollable

This item should be fixed now.

>I am still seeing this darker line under Analytics

The highlighted darker line is coming from the JSON settings. I just opened a PR that should address this issue (see - https://github.ibm.com/webstandards/ibm-dotcom-translations/pull/383). Once it has been merged, you should see this darker line again.

>I also noticed that we are not breaking to the hamburger until around 800px. I think we should break at 959px instead since the menu items are unusable at that size.

I've discussed this request with @annawen1 and we agree that it's difficult to scope this breakpoint change to just the Cloud Masthead as we're currently using the same components and extending them will create more technical debts. Our recommendation is to adjust the visibility of the mobile menu for both the global and cloud versions, if possible.

>If the left side menu tabs do not fit vertically in the browser window I would expect that we have scrolling only for that side of the mega menu rather than making both sides scroll. Once the content does not fit I expect 2 scroll bars one for left and one for right side content as needed. 

The left side menu tabs area is just a div and not an iFrame that could easily have the vertical scrolling behavior. Also, I believe this was discussed at some point within the design team as well and there's some concern regarding the UX. I would come back to this in a different design enhancement ticket.

>I think it would make sense if the View all products keeps the blue styling it had at larger breakpoints even at mobile size. Will confirm this with @shixiedesign and @oliviaflory

With this particular change, I think we might need to get a sign-off first before bringing it into dev. Similar to the item above, I would capture this in a different design enhancement ticket.

> [Left Nav: Web Components]: L1 items are visible when on L2 menu items

Known issue that @annawen1 is currently looking into. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6209

**Comment by ltitle** [comment]
@proeung Thank you for the updates above. I will connect with the team on the scrolling issue because that is concerning to me. A couple other things that do not seem to be resolved:

**Item layout in the mega menu** 
Some pages are still not filling out 3 across before wrapping to a new line

image
image
image

**Height of mega menu for Solutions is changing with each selected topic**
Should maintain the height of the longest content set so it is not jumping up and down while using the menu

image
image

**Comment by kjwhitla** [comment]
Updates match the intended layout and design interaction. No issues from my end.

**Comment by proeung** [comment]
@ltitle I just pushed up some changes based on what we discussed on our call today. We're using flexbox to align the items within each category and I just adjusted the hover background to fill in the box as well. The preview is still building, but you should be able to see the changes soon (~ 30 mins or so). https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5922/index.html?path=/story/components-cloud-masthead--default

I just opened a ticket (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6240) for the remaining design items that are marked as follow-up enhancements. Feel free to add, if I'm missing anything.

**Comment by ariellalgilmore** [comment]
Looking great!!!! small thing: it looks like in react the color did not change for the profile icon 
Screen Shot 2021-06-02 at 1 21 35 PM
vs web components
Screen Shot 2021-06-02 at 1 21 56 PM

**Comment by proeung** [comment]
> Looking great!!!! small thing: it looks like in react the color did not change for the profile icon

@ariellalgilmore Thanks for the review! I've adjusted the color within the `_masthead-search.scss`, which I assumed is a shared stylesheet for both React and WC. Are we using a different class for the react version?

**Review by ariellalgilmore — APPROVED** [review]
LGTM!

**Comment by shixiedesign** [comment]
@ltitle lemme know whenever you think this is good to go and I can remove the "change requested"!

**Comment by jeffchew** [comment]
@shixiedesign would you be able to approve percy changes if they look ok to you?

**Comment by ltitle** [comment]
@proeung @shixiedesign Looks good to go!

**Review by shixiedesign — APPROVED** [review]
🔥

**Comment by oliviaflory** [comment]
Hi @proeung ! I'm not sure if you've seen this, there's something between the search icon and My IBM that the user can select / tab to.

CLOUD-tab-Jul-16-2021 14-38-00

**Pull request merged**
